File: sha9001.lua
function rot(value,distance,key) if value==nil then error("ROT input is null") end; if value:find("%z") then error("ROT input contains NUL") end; d=((distance%26)+26)%26; return (value:gsub("[A-Za-z]",function(c)o=c:byte();b=o<=90 and 65 or 97;return string.char(b+(o-b+d)%26)end)) end
function rotn(value,distance,key) if type(distance)~="number" or distance%1~=0 then error("ROT distance must be an integer") end; if type(key)~="string" then error("ROT key must be bytes") end; return rot(value,distance,key) end
function rot13(value,key)return rotn(value,13,key) end
function ebg13(value,key)return rotn(value,-13,key) end
function sha9001(data) sha1=require("sha1");digest=data;for _=1,9001 do digest=sha1.binary(digest) end;return digest end

File: sha9001.js
'use strict'; const crypto=require('crypto');
function sha9001(data){let d=Buffer.from(data);for(let i=0;i<9001;i++)d=crypto.createHash('sha1').update(d).digest();return d}
function sanitize(v){if(v===null||v===undefined)throw new TypeError('ROT input is null');if(v.includes('\0'))throw new TypeError('ROT input contains a NUL character');return v}
function rot(value,distance){const d=((distance%26)+26)%26;return [...value].map(c=>{const o=c.charCodeAt(0),b=o>=65&&o<=90?65:o>=97&&o<=122?97:null;return b===null?c:String.fromCharCode(b+(o-b+d)%26)}).join('')}
function rotn(value,distance,key){value=sanitize(value);if(!Number.isInteger(distance))throw new TypeError('ROT distance must be an integer');if(!Buffer.isBuffer(key)&&!(key instanceof Uint8Array))throw new TypeError('ROT key must be bytes');return rot(value,distance)}
const rot13=(v,k)=>rotn(v,13,k),ebg13=(v,k)=>rotn(v,-13,k); module.exports={sha9001,rot13,ebg13,rotn}
